Edits to queries article for 3.0 (#3235)
apollographql/apollo-android: docs/source/essentials/00-queries.mdx

---
title: Queries in Apollo Android
sidebar_title: Queries
---

import {MultiCodeBlock} from 'gatsby-theme-apollo-docs';

Fetching data in a predictable, type-safe way is one of the core features of Apollo Android. In this guide, you'll learn how to query a GraphQL endpoint and use the result in your application.

## Prerequisites

> This page assumes some familiarity with building GraphQL queries. For a refresher, we recommend [reading this guide](http://graphql.org/learn/queries/) and practicing [running queries in Apollo Sandbox](https://studio.apollographql.com/sandbox/explorer?endpoint=https%3A%2F%2Fswapi-graphql.netlify.app%2F.netlify%2Ffunctions%2Findex&explorerURLState=N4IgJg9gxgrgtgUwHYBcQC4QEcYIE4CeABAOIIoBiAlgDZwDORwAOkkUQIY03V2Mtt2RAGa0GTVkKEoqKGgklCAvopVIlIADQgAbhzxUOAI3n0MIEEqA).

Reviewer: 👍

> Because Apollo Android uses standard GraphQL syntax, anything query you can run in Sandbox can also be put into `.graphql` files in your project.

This page also assumes that you've already set up Apollo Android for your Android/Kotlin application. For help with setup, see the [getting started guide](../#get-started).

## Defining

> In Apollo Android, each query you execute is represented as an instance of a generated class that conforms to the `Query` protocol. To generate these classes, we first need to define the GraphQL operations we want to execute.

Author: Do we have javadoc or similar for the library? If not, we can just link to wherever in the GH source the `Query` protocol is defined

Reviewer: Javadoc has been on my todo list for a while. It'd be super useful for cases like this. Do you have a preferred option how to host them? I think Dokka (the documentation engine for Kotlin) can export Github markdown (example [here](https://square.github.io/okhttp/4.x/okhttp/okhttp3/)) so we could add it as a separate section in the gatsby site. This would only contain the latest API version but maybe that's ok?

Reviewer: With the permalink to Github (and replacing some borrowed Swifties by Kotlin interfaces 🙈 )

```suggestion
In Apollo Android, each query you execute is represented as an instance of a generated class that conforms to the [`Query` interface](https://github.com/apollographql/apollo-android/blob/9c998e47f698bdbe5db6b6d6965dbc34183d507e/apollo-api/src/commonMain/kotlin/com/apollographql/apollo3/api/Query.kt#L6). To generate these classes, we first need to define the GraphQL operations we want to execute.
```

Let's say we define a GraphQL query named `HeroQuery`, like so:

> ```graphql:title=hero-query.graphql

Author: Not sure what the filename convention should be here, happy to adjust

Reviewer: We don't have really strong suggestions for naming files/operations (but I think we should, same for fragments). For the time being, I'd do:

```suggestion
```graphql:title=HeroQuery.graphql
```

> ```graphql:title=hero-query.graphql

Reviewer: We don't have really strong suggestions for naming (but I think we should, same for fragments). For the time being, I'd do:

```suggestion
```graphql:title=HeroQuery.graphql
```

}
```

If we define this query in its own `.graphql` file, Apollo Android will generate a corresponding `HeroQuery` class that we can then use to [execute the query](#executing).

> To generate classes, Apollo Android requires your server's schema and a set of `.graphql` files (one for each operation in your application). It uses these to generate code you can use to execute queries and access typed results.

Author: Does it actually have to be one operation per file? Or can you define multiple operations in a single file?

Reviewer: It can be multiple operations in a single file. In the end, all the `.graphql` files are concatenated together.

> All `.graphql` files in your project (or the subset you specify as input to `apollo-codegen` if you customize your code generation build phase script) are combined and treated as _one_ GraphQL document. This means that fragments defined in one `.graphql` file are available across _all_ `.graphql` files. However, it _also_ means that operation and fragment names must be unique (and validation errors will occur if they aren't).

## Executing

To execute the query we [just defined](#defining), we pass a `HeroQuery` instance to the `query` method of `ApolloClient`, like so:

> ```kotlin{2}

Reviewer: 👀  What does `{2}` do ?

Author: Slightly highlights line 2 ✨

Reviewer: ![giphy](https://user-images.githubusercontent.com/3974977/125664735-c2d3e610-2bd0-4a13-801d-e7899a2cfcc0.gif)
